File: src/Hardware_Detector.cpp
#include "Hardware_Detector.hpp"
#include <iostream>
#include <cstring>
#include <thread>
#include <fstream>
#include <sstream>

// ---------------------------------------------------------------------------
// Platform-specific CPU model string
// ---------------------------------------------------------------------------
#if defined(__APPLE__)

#include <sys/sysctl.h>

static std::string Get_CPU_Model()
{
    char buf[128];
    size_t len = sizeof(buf);
    if (sysctlbyname("machdep.cpu.brand_string", buf, &len, nullptr, 0) == 0)
        return buf;
    return "Unknown Apple CPU";
}

#elif defined(__linux__)

static std::string Get_CPU_Model()
{
    std::ifstream cpuinfo("/proc/cpuinfo");
    if (!cpuinfo.is_open())
        return "Unknown";

    std::string line;
    while (std::getline(cpuinfo, line))
    {
        if (line.find("model name") != std::string::npos)
        {
            auto pos = line.find(':');
            if (pos != std::string::npos)
            {
                // Trim leading whitespace
                std::string name = line.substr(pos + 1);
                name.erase(0, name.find_first_not_of(" \t"));
                return name;
            }
        }
    }
    return "Unknown x86 CPU";
}

#else

static std::string Get_CPU_Model()
{
    return "Unknown CPU";
}

#endif

// ---------------------------------------------------------------------------
// Platform-specific core counts
// ---------------------------------------------------------------------------
#if defined(__APPLE__)

static unsigned Get_Performance_Cores()
{
#ifdef __aarch64__
    int cores = 0;
    size_t size = sizeof(cores);
    sysctlbyname("hw.perflevel0.physicalcpu", &cores, &size, nullptr, 0);
    return static_cast<unsigned>(cores > 0 ? cores : 0);
#else
    return 0;   // Intel Macs don't expose P/E count via sysctl
#endif
}

static unsigned Get_Efficiency_Cores()
{
#ifdef __aarch64__
    int cores = 0;
    size_t size = sizeof(cores);
    sysctlbyname("hw.perflevel1.physicalcpu", &cores, &size, nullptr, 0);
    return static_cast<unsigned>(cores > 0 ? cores : 0);
#else
    return 0;
#endif
}

#else

static unsigned Get_Performance_Cores() { return 0; }
static unsigned Get_Efficiency_Cores()  { return 0; }

#endif

// ---------------------------------------------------------------------------
// SIMD feature detection
// ---------------------------------------------------------------------------
#if defined(__x86_64__) || defined(_M_X64)

#include <cpuid.h>

static bool Supports_AVX()
{
    unsigned int eax, ebx, ecx, edx;
    __cpuid(1, eax, ebx, ecx, edx);
    return (ecx & (1U << 28)) != 0;
}

static bool Supports_AVX2()
{
    unsigned int eax, ebx, ecx, edx;
    __cpuid_count(7, 0, eax, ebx, ecx, edx);
    return (ebx & (1U << 5)) != 0;
}

static bool Supports_NEON() { return false; }

#elif defined(__aarch64__)

static bool Supports_AVX()   { return false; }
static bool Supports_AVX2()  { return false; }

#if defined(__APPLE__)
// NEON is baseline on Apple Silicon
static bool Supports_NEON()  { return true; }
#elif defined(__linux__)
static bool Supports_NEON()
{
    std::ifstream cpuinfo("/proc/cpuinfo");
    if (!cpuinfo.is_open())
        return false;
    std::string line;
    while (std::getline(cpuinfo, line))
    {
        if (line.find("Features") != std::string::npos)
        {
            if (line.find("neon") != std::string::npos ||
                line.find("asimd") != std::string::npos)
                return true;
        }
    }
    return false;
}
#else
static bool Supports_NEON()  { return false; }
#endif

#else

static bool Supports_AVX()   { return false; }
static bool Supports_AVX2()  { return false; }
static bool Supports_NEON()  { return false; }

#endif

// ---------------------------------------------------------------------------
// CPUInfo::CPU_Detect
// ---------------------------------------------------------------------------
void CPUInfo::CPU_Detect()
{
    model        = Get_CPU_Model();
    logicalCores = std::thread::hardware_concurrency();
    perfCores    = Get_Performance_Cores();
    effiCores    = Get_Efficiency_Cores();
    hasAVX       = Supports_AVX();
    hasAVX2      = Supports_AVX2();
    hasNEON      = Supports_NEON();
}

File: rust/src/hardware_detector.rs
#[cfg(target_os = "linux")]
use std::fs;
#[cfg(target_os = "macos")]
use std::process::Command;

pub struct CpuInfo {
    pub model: String,
    pub logical_cores: usize,
    pub has_avx: bool,
    pub has_avx2: bool,
}

impl CpuInfo {
    pub fn detect() -> Self {
        Self {
            model: detect_cpu_model(),
            logical_cores: std::thread::available_parallelism()
                .map(usize::from)
                .unwrap_or(0),
            has_avx: detect_avx(),
            has_avx2: detect_avx2(),
        }
    }
}

fn detect_cpu_model() -> String {
    detect_cpu_model_impl()
        .filter(|model| !model.is_empty())
        .unwrap_or_else(|| "Unknown".to_string())
}

#[cfg(target_os = "macos")]
fn detect_cpu_model_impl() -> Option<String> {
    run_command("sysctl", &["-n", "machdep.cpu.brand_string"])
        .or_else(|| run_command("sysctl", &["-n", "hw.model"]))
        .or_else(|| match std::env::consts::ARCH {
            "aarch64" | "arm64" => Some("Apple Silicon".to_string()),
            arch => Some(format!("macOS ({arch})")),
        })
}

#[cfg(target_os = "linux")]
fn detect_cpu_model_impl() -> Option<String> {
    let cpuinfo = fs::read_to_string("/proc/cpuinfo").ok()?;
    cpuinfo.lines().find_map(|line| {
        let (key, value) = line.split_once(':')?;
        (key.trim() == "model name").then(|| value.trim().to_string())
    })
}

#[cfg(target_os = "windows")]
fn detect_cpu_model_impl() -> Option<String> {
    std::env::var("PROCESSOR_IDENTIFIER").ok()
}

#[cfg(not(any(target_os = "macos", target_os = "linux", target_os = "windows")))]
fn detect_cpu_model_impl() -> Option<String> {
    None
}

#[cfg(target_os = "macos")]
fn run_command(command: &str, args: &[&str]) -> Option<String> {
    let output = Command::new(command).args(args).output().ok()?;
    if !output.status.success() {
        return None;
    }

    let value = String::from_utf8(output.stdout).ok()?;
    let value = value.trim();
    (!value.is_empty()).then(|| value.to_string())
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
fn detect_avx() -> bool {
    std::arch::is_x86_feature_detected!("avx")
}

#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
fn detect_avx() -> bool {
    false
}

#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
fn detect_avx2() -> bool {
    std::arch::is_x86_feature_detected!("avx2")
}

#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
fn detect_avx2() -> bool {
    false
}
